File: android/app/src/main/java/com/ncocaptain/simpledinners/ShareRecipeExtractorPlugin.java
package com.ncocaptain.simpledinners;

import android.annotation.SuppressLint;
import android.os.Handler;
import android.os.Looper;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "ShareRecipeExtractor")
public class ShareRecipeExtractorPlugin extends Plugin {

    private static final long INSTAGRAM_RENDER_DELAY_MS = 8000L;

    @SuppressLint("SetJavaScriptEnabled")
    @PluginMethod
    public void extractJsonLd(PluginCall call) {
        String url = call.getString("url");

        if (url == null || url.trim().isEmpty()) {
            call.reject("URL required");
            return;
        }

        getActivity().runOnUiThread(() -> {
            WebView webView = new WebView(getContext());

            WebSettings settings = webView.getSettings();
            settings.setJavaScriptEnabled(true);
            settings.setDomStorageEnabled(true);
            settings.setUserAgentString(
                "Mozilla/5.0 (Linux; Android 13; Mobile) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Mobile Safari/537.36"
            );

            webView.setWebViewClient(new WebViewClient() {
                private boolean extracted = false;

                @Override
                public void onPageFinished(WebView view, String finishedUrl) {
                    if (extracted) return;
                    extracted = true;

                    new Handler(Looper.getMainLooper()).postDelayed(() -> {
                        String script =
                            "(function() {" +
                            "  function findRecipe(value) {" +
                            "    if (!value) return null;" +
                            "    if (Array.isArray(value)) {" +
                            "      for (const item of value) {" +
                            "        const found = findRecipe(item);" +
                            "        if (found) return found;" +
                            "      }" +
                            "      return null;" +
                            "    }" +
                            "    if (typeof value !== 'object') return null;" +
                            "    const type = value['@type'];" +
                            "    const isRecipe = type === 'Recipe' || (Array.isArray(type) && type.includes('Recipe'));" +
                            "    const looksLikeRecipe = value.recipeIngredient || value.recipeInstructions || value.cookTime || value.prepTime;" +
                            "    if (isRecipe || looksLikeRecipe) return value;" +
                            "    for (const key of Object.keys(value)) {" +
                            "      const found = findRecipe(value[key]);" +
                            "      if (found) return found;" +
                            "    }" +
                            "    return null;" +
                            "  }" +
                            "  const scripts = Array.from(document.querySelectorAll('script[type=\"application/ld+json\"]'));" +
                            "  for (const s of scripts) {" +
                            "    try {" +
                            "      const parsed = JSON.parse(s.textContent || '');" +
                            "      const recipe = findRecipe(parsed);" +
                            "      if (recipe) return JSON.stringify(recipe);" +
                            "    } catch (e) {}" +
                            "  }" +
                            "  return '';" +
                            "})();";

                        view.evaluateJavascript(script, value -> {
                            String cleanedJsonLd = value == null ? "" : value;

                            if (cleanedJsonLd.startsWith("\"") && cleanedJsonLd.endsWith("\"")) {
                                cleanedJsonLd = cleanedJsonLd.substring(1, cleanedJsonLd.length() - 1)
                                    .replace("\\n", "\n")
                                    .replace("\\\"", "\"")
                                    .replace("\\\\", "\\");
                            }

                            JSObject result = new JSObject();
                            result.put("url", finishedUrl);
                            result.put("jsonLd", cleanedJsonLd);
                            result.put("length", cleanedJsonLd.length());

                            call.resolve(result);

                            try {
                                webView.destroy();
                            } catch (Exception ignored) {
                            }
                        });
                    }, 5000);
                }
            });

            webView.loadUrl(url.trim());
        });
    }

    @SuppressLint("SetJavaScriptEnabled")
    @PluginMethod
    public void extractInstagramCaption(PluginCall call) {
        String url = call.getString("url");

        if (url == null || url.trim().isEmpty()) {
            call.reject("URL required");
            return;
        }

        String trimmedUrl = url.trim();

        if (!isInstagramUrl(trimmedUrl)) {
            call.reject("Instagram URL required");
            return;
        }

        getActivity().runOnUiThread(() -> {
            WebView webView = new WebView(getContext());

            WebSettings settings = webView.getSettings();
            settings.setJavaScriptEnabled(true);
            settings.setDomStorageEnabled(true);
            settings.setUserAgentString(
                "Mozilla/5.0 (Linux; Android 13; Mobile) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Mobile Safari/537.36"
            );

            webView.setWebViewClient(new WebViewClient() {
                private boolean extractionScheduled = false;

                @Override
                public boolean shouldOverrideUrlLoading(
                    WebView view,
                    android.webkit.WebResourceRequest request
                ) {
                    String scheme = request.getUrl().getScheme();

                    if (
                        "http".equalsIgnoreCase(scheme) ||
                        "https".equalsIgnoreCase(scheme)
                    ) {
                        return false;
                    }

                    // Keep Instagram from replacing the rendered HTTPS page
                    // with an instagram:// app deep link.
                    return true;
                }

                @Override
                public void onPageStarted(
                    WebView view,
                    String startedUrl,
                    android.graphics.Bitmap favicon
                ) {
                    if (extractionScheduled) return;
                    extractionScheduled = true;

                    new Handler(Looper.getMainLooper()).postDelayed(() -> {
                        String script =
                            "(function() {" +
                            "  const getMeta = (selector) =>" +
                            "    document.querySelector(selector)?.getAttribute('content') || '';" +
                            "  return JSON.stringify({" +
                            "    url: location.href || ''," +
                            "    captionText: getMeta('meta[name=\"description\"]')," +
                            "    photoUrl: getMeta('meta[property=\"og:image\"]')," +
                            "    ogTitle: getMeta('meta[property=\"og:title\"]')" +
                            "  });" +
                            "})();";

                        view.evaluateJavascript(script, value -> {
                            try {
                                String rawJson = decodeJavascriptString(value);
                                org.json.JSONObject payload = rawJson.isEmpty()
                                    ? new org.json.JSONObject()
                                    : new org.json.JSONObject(rawJson);

                                String captionText = payload.optString("captionText", "").trim();
                                String photoUrl = payload.optString("photoUrl", "").trim();
                                String ogTitle = payload.optString("ogTitle", "").trim();
                                String finalUrl = payload.optString("url", "").trim();

                                if (finalUrl.isEmpty()) {
                                    finalUrl = view.getUrl() == null ? trimmedUrl : view.getUrl();
                                }

                                JSObject result = new JSObject();
                                result.put("url", finalUrl);
                                result.put("captionText", captionText);
                                result.put("photoUrl", photoUrl);
                                result.put("ogTitle", ogTitle);
                                result.put("length", captionText.length());

                                call.resolve(result);
                            } catch (Exception error) {
                                call.reject("Could not read Instagram page metadata.");
                            } finally {
                                try {
                                    webView.destroy();
                                } catch (Exception ignored) {
                                }
                            }
                        });
                    }, INSTAGRAM_RENDER_DELAY_MS);
                }
            });

            webView.loadUrl(trimmedUrl);
        });
    }

    private boolean isInstagramUrl(String rawUrl) {
        try {
            android.net.Uri uri = android.net.Uri.parse(rawUrl);
            String scheme = uri.getScheme();
            String host = uri.getHost();

            if (
                scheme == null ||
                host == null ||
                !("http".equalsIgnoreCase(scheme) || "https".equalsIgnoreCase(scheme))
            ) {
                return false;
            }

            String normalizedHost = host.toLowerCase();

            return normalizedHost.equals("instagram.com") ||
                normalizedHost.endsWith(".instagram.com");
        } catch (Exception ignored) {
            return false;
        }
    }

    private String decodeJavascriptString(String value) {
        if (value == null || value.trim().isEmpty() || "null".equals(value)) {
            return "";
        }

        try {
            Object decoded = new org.json.JSONTokener(value).nextValue();
            return decoded instanceof String ? (String) decoded : "";
        } catch (Exception ignored) {
            return "";
        }
    }
}

File: ios/App/App/ShareRecipeExtractorPlugin.swift
import Capacitor
import WebKit

@objc(ShareRecipeExtractorPlugin)
public class ShareRecipeExtractorPlugin: CAPPlugin {
    private var extractorWebView: WKWebView?
    private var currentCall: CAPPluginCall?

    private var instagramWebView: WKWebView?
    private var instagramCall: CAPPluginCall?
    private var instagramExtractionScheduled = false


    @objc func extractJsonLd(_ call: CAPPluginCall) {
        guard let urlString = call.getString("url"),
              let url = URL(string: urlString) else {
            call.reject("URL required")
            return
        }

        currentCall = call

        DispatchQueue.main.async {
            let config = WKWebViewConfiguration()
            let hiddenWebView = WKWebView(frame: .zero, configuration: config)

            self.extractorWebView = hiddenWebView
            hiddenWebView.navigationDelegate = self
            hiddenWebView.load(URLRequest(url: url))
        }
    }

    @objc func extractInstagramCaption(_ call: CAPPluginCall) {
        guard
            let urlString = call.getString("url"),
            let url = URL(string: urlString),
            isInstagramURL(url)
        else {
            call.reject("Instagram URL required")
            return
        }

        instagramCall = call
        instagramExtractionScheduled = false

        DispatchQueue.main.async {
            let config = WKWebViewConfiguration()
            let hiddenWebView = WKWebView(
                frame: .zero,
                configuration: config
            )

            hiddenWebView.customUserAgent =
                "Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) " +
                "AppleWebKit/605.1.15 (KHTML, like Gecko) " +
                "Version/17.0 Mobile/15E148 Safari/604.1"

            self.instagramWebView = hiddenWebView
            hiddenWebView.navigationDelegate = self
            hiddenWebView.load(URLRequest(url: url))
        }
    }

    private func isInstagramURL(_ url: URL) -> Bool {
        guard
            let scheme = url.scheme?.lowercased(),
            scheme == "http" || scheme == "https",
            let host = url.host?.lowercased()
        else {
            return false
        }

        return host == "instagram.com" ||
            host.hasSuffix(".instagram.com")
    }

    private func finishInstagramExtraction(
        captionText: String,
        photoUrl: String,
        ogTitle: String,
        url: String
    ) {
        instagramCall?.resolve([
            "url": url,
            "captionText": captionText,
            "photoUrl": photoUrl,
            "ogTitle": ogTitle,
            "length": captionText.count
        ])

        instagramCall = nil
        instagramExtractionScheduled = false
        instagramWebView?.navigationDelegate = nil
        instagramWebView = nil
    }

    private func finish(jsonLd: String, url: String) {
        currentCall?.resolve([
            "url": url,
            "jsonLd": jsonLd,
            "length": jsonLd.count
        ])

        currentCall = nil
        extractorWebView?.navigationDelegate = nil
        extractorWebView = nil
    }
}

extension ShareRecipeExtractorPlugin: WKNavigationDelegate {
    public func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
        if webView === instagramWebView {
            if instagramExtractionScheduled {
                return
            }

            instagramExtractionScheduled = true

            let instagramScript = """
            (function() {
              const getMeta = (selector) =>
                document.querySelector(selector)?.getAttribute('content') || '';

              return JSON.stringify({
                url: location.href || '',
                captionText: getMeta('meta[name="description"]'),
                photoUrl: getMeta('meta[property="og:image"]'),
                ogTitle: getMeta('meta[property="og:title"]')
              });
            })();
            """

            DispatchQueue.main.asyncAfter(deadline: .now() + 8.0) {
                webView.evaluateJavaScript(instagramScript) { result, _ in
                    let rawJson = result as? String ?? ""

                    var captionText = ""
                    var photoUrl = ""
                    var ogTitle = ""
                    var finalUrl =
                        webView.url?.absoluteString ?? ""

                    if
                        let jsonData = rawJson.data(using: .utf8),
                        let payload =
                            try? JSONSerialization.jsonObject(
                                with: jsonData
                            ) as? [String: Any]
                    {
                        captionText =
                            (payload["captionText"] as? String ?? "")
                            .trimmingCharacters(
                                in: .whitespacesAndNewlines
                            )

                        photoUrl =
                            (payload["photoUrl"] as? String ?? "")
                            .trimmingCharacters(
                                in: .whitespacesAndNewlines
                            )

                        ogTitle =
                            (payload["ogTitle"] as? String ?? "")
                            .trimmingCharacters(
                                in: .whitespacesAndNewlines
                            )

                        let payloadUrl =
                            (payload["url"] as? String ?? "")
                            .trimmingCharacters(
                                in: .whitespacesAndNewlines
                            )

                        if !payloadUrl.isEmpty {
                            finalUrl = payloadUrl
                        }
                    }

                    self.finishInstagramExtraction(
                        captionText: captionText,
                        photoUrl: photoUrl,
                        ogTitle: ogTitle,
                        url: finalUrl
                    )
                }
            }

            return
        }

        let script = """
        (function() {
          function findRecipe(value) {
            if (!value) return null;
            if (Array.isArray(value)) {
              for (const item of value) {
                const found = findRecipe(item);
                if (found) return found;
              }
              return null;
            }
            if (typeof value !== 'object') return null;
            const type = value['@type'];
            const isRecipe = type === 'Recipe' || (Array.isArray(type) && type.includes('Recipe'));
            const looksLikeRecipe = value.recipeIngredient || value.recipeInstructions || value.cookTime || value.prepTime;
            if (isRecipe || looksLikeRecipe) return value;
            for (const key of Object.keys(value)) {
              const found = findRecipe(value[key]);
              if (found) return found;
            }
            return null;
          }

          const scripts = Array.from(document.querySelectorAll('script[type="application/ld+json"]'));

          for (const s of scripts) {
            try {
              const parsed = JSON.parse(s.textContent || '');
              const recipe = findRecipe(parsed);
              if (recipe) return JSON.stringify(recipe);
            } catch (e) {}
          }

          return '';
        })();
        """

        DispatchQueue.main.asyncAfter(deadline: .now() + 2.0) {
            webView.evaluateJavaScript(script) { result, _ in
                let jsonLd = result as? String ?? ""
                self.finish(jsonLd: jsonLd, url: webView.url?.absoluteString ?? "")
            }
        }
    }
}
